src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReaderStream.cs

```diff
@@ -0,0 +1,811 @@
+                        // since the caller should already have checked this.
+                        throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReaderStreamBuffer_ReadWithLength));
+                    }
+                    else
+                    {
+                        this.underlyingStreamExhausted = this.BatchBuffer.RefillFrom(this.multipartMixedBatchInputContext.Stream, /*preserveFrom*/ ODataBatchReaderStreamBuffer.BufferLength);
+                    }
+                }
+            }
+
+            // return the number of bytes that were read
+            return count - remainingNumberOfBytesToRead;
+        }
+
+        /// <summary>
+        /// Reads the headers of a part.
+        /// </summary>
+        /// <param name="contentId">Content-ID read from changeset header, null if changeset part detected</param>
+        /// <param name="dependsOnIds">DependsOn-IDs read from request header, null if the header is not present.</param>
+        /// <returns>true if the start of a changeset part was detected; otherwise false.</returns>
+        internal bool ProcessPartHeader(out string contentId, out string dependsOnIds)
+        {
+            Debug.Assert(this.batchEncoding != null, "Batch encoding should have been established on first call to SkipToBoundary.");
+
+            bool isChangeSetPart;
+            ODataBatchOperationHeaders headers = this.ReadPartHeaders(out isChangeSetPart);
+
+            contentId = null;
+            dependsOnIds = null;
+
+            if (isChangeSetPart)
+            {
+                // determine the changeset boundary and the changeset encoding from the content type header
+                this.DetermineChangesetBoundaryAndEncoding(headers[ODataConstants.ContentTypeHeader]);
+
+                if (this.changesetEncoding == null)
+                {
+                    // NOTE: No changeset encoding was specified in the changeset's content type header.
+                    //       Determine the changeset encoding from the first bytes in the changeset.
+                    // NOTE: We do not have to skip over the potential preamble of the encoding
+                    //       because the batch reader will skip over everything (incl. the preamble)
+                    //       until it finds the first changeset (or batch) boundary
+                    this.changesetEncoding = this.DetectEncoding(this.multipartMixedBatchInputContext.Stream);
+                }
+
+                // Verify that we only allow single byte encodings and UTF-8 for now.
+                ReaderValidationUtils.ValidateEncodingSupportedInBatch(this.changesetEncoding);
+            }
+            else if (this.ChangeSetBoundary != null)
+            {
+                headers.TryGetValue(ODataConstants.ContentIdHeader, out contentId);
+                headers.TryGetValue(ODataConstants.DependsOnIdsHeader, out dependsOnIds);
+            }

```

Comment:
See other comments. I wasn't suggesting that we introduce a specific way to extend multipart/mixed with dependencyids. The protocol already defines that, for the multipartmixed format, each statement is implicitly dependent upon the previous one, so we can just populate the dependencyIds list with the list of all previous statements in the same changeset (or at the top level, for requests outside of a changeset).

Reply:
rolled back the previous implementation. 

---
In reply to: [156803180](https://github.com/OData/odata.net/pull/992#discussion_r156803180) [](ancestors = 156803180)